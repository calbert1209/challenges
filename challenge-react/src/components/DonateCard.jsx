import React, { useState } from 'react';
import styled from 'styled-components';
import Button from './Button';
import DonateForm from './DonateForm';
import Modal from './Modal';

const Card = styled.div`
  height: 270px;
  border: 1px solid #ccc;
  box-sizing: border-box;
  display: flex;
  border-radius: 3px;
  position: relative;
  box-shadow: 2px 2px 4px 0px #e2e2e2;
  flex-direction: column;
`;

const CartMedia = styled.img`
  height: calc(100% - 70px);
  height: 100%;
  width: 100%;
  object-fit: cover;
  object-position: top;
`;

const CardFooter = styled.div`
  display: flex;
  justify-content: space-between;
  flex-grow: 1;
  margin: 20px;
  align-items: center;
`;

const DonateCard = (props) => {
  const [formVisible, setFormVisible] = useState(false);
  const handleSubmit = (e) => {
    e.preventDefault();
    props.onSubmit(
      props.item.id,
      parseInt(e.target.payment.value),
      props.item.currency
    );
    e.target.reset();
    toggleDonateForm();
  };

  const toggleDonateForm = () => {
    setFormVisible(!formVisible);
  };
  return (
    <Card className={props.className}>
      <Modal visible={formVisible} onClose={toggleDonateForm}>
        <DonateForm onSubmit={handleSubmit} currency={props.item.currency} />
      </Modal>
      <div style={{ overflow: 'hidden' }}>
        <CartMedia src={`/images/${props.item.image}`} />
      </div>
      <CardFooter>
        <p style={{ margin: 0 }}>{props.item.name}</p>

        <Button id={'donate-btn'} onClick={toggleDonateForm}>
          Donate
        </Button>
      </CardFooter>
    </Card>
  );
};

export default DonateCard;
