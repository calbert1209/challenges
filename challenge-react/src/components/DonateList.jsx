import React from 'react';
import styled from 'styled-components';
import DonateCard from './DonateCard';

const CardWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  @media (min-width: 1024px) {
    margin: 0 50px;
  }
`;

const DonateCardItem = styled(DonateCard)`
  margin: 20px;
  width: 100%;
  @media (min-width: 1024px) {
    width: calc(50% - 40px);
  }
`;

const DonateList = ({ charities, handlePay }) => {
  const handleSubmit = (id, value, currency) => {
    handlePay(id, value, currency);
  };
  return (
    <CardWrapper>
      {charities.map((item) => (
        <DonateCardItem item={item} key={item.id} onSubmit={handleSubmit} />
      ))}
    </CardWrapper>
  );
};

export default DonateList;
