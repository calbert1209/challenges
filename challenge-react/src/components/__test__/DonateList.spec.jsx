import React from 'react';
import TestRenderer from 'react-test-renderer';
import DonateCard from '../DonateCard';
import DonateList from '../DonateList';
describe('DonateList', () => {
  it('render correctly', () => {
    const mockItem = { id: '1', image: 'test.jpg', name: 'test' };
    const testRenderer = TestRenderer.create(
      <DonateList charities={[mockItem]} handlePay={jest.fn()} />
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
  it('on pay', () => {
    const mockItem = {
      id: '1',
      image: 'test.jpg',
      name: 'test',
      currency: 'THB',
    };
    const mockValue = 50;
    const mockHandlePay = jest.fn();
    const testRenderer = TestRenderer.create(
      <DonateList charities={[mockItem]} handlePay={mockHandlePay} />
    );
    const testInstance = testRenderer.root;
    testInstance
      .findAllByType(DonateCard)[0]
      .props.onSubmit(mockItem.id, mockValue, mockItem.currency);

    expect(mockHandlePay).toHaveBeenCalledWith(
      mockItem.id,
      mockValue,
      mockItem.currency
    );
  });
});
