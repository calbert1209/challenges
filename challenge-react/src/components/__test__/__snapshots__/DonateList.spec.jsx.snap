// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonateList render correctly 1`] = `
<div
  className="sc-jrAGrp dVWBci"
>
  <div
    className="sc-iBPRYJ hyDFeT sc-kEjbxe etBImo"
  >
    <div
      className="sc-jSgupP iiJNAZ"
    >
      <div
        className="sc-gKsewC hIxykI"
        onClick={[Function]}
      />
      <div
        className="sc-gsTCUz dCGfcy"
      >
        <h3>
          Select the amiunt to donate (
          )
        </h3>
        <form
          className="sc-dlfnbm lgmEEb"
          onSubmit={[Function]}
        >
          <div>
            <label
              className="sc-eCssSg iNcjtx"
            >
              <input
                name="payment"
                required={true}
                type="radio"
                value={10}
              />
              10
            </label>
            <label
              className="sc-eCssSg iNcjtx"
            >
              <input
                name="payment"
                type="radio"
                value={20}
              />
              20
            </label>
            <label
              className="sc-eCssSg iNcjtx"
            >
              <input
                name="payment"
                type="radio"
                value={50}
              />
              50
            </label>
            <label
              className="sc-eCssSg iNcjtx"
            >
              <input
                name="payment"
                type="radio"
                value={100}
              />
              100
            </label>
            <label
              className="sc-eCssSg iNcjtx"
            >
              <input
                name="payment"
                type="radio"
                value={500}
              />
              500
            </label>
          </div>
          <div
            className="sc-hKgILt kJXgYw"
          >
            <button
              className="sc-bdfBwQ eCLJup"
              type="submit"
            >
              Pay
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      style={
        Object {
          "overflow": "hidden",
        }
      }
    >
      <img
        className="sc-fubCfw diUEuu"
        src="/images/test.jpg"
      />
    </div>
    <div
      className="sc-pFZIQ kgmWnS"
    >
      <p
        style={
          Object {
            "margin": 0,
          }
        }
      >
        test
      </p>
      <button
        className="sc-bdfBwQ eCLJup"
        id="donate-btn"
        onClick={[Function]}
      >
        Donate
      </button>
    </div>
  </div>
</div>
`;
